pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		rg.w.Writefmtln(" }")
+	default:
+		rg.w.Writefmtln(",")
+		rg.w.Writefmtln("                Aliases =")
+		rg.w.Writefmtln("                {")
+		for i, alias := range rg.res.info.Aliases {
+			rg.w.Writefmt("                    ")
+			rg.generateAlias(alias)
+			if i < len(rg.res.info.Aliases)-1 {
+				rg.w.Writefmtln(",")
+			}
+		}
+		rg.w.Writefmtln("                }")
+	}
+
+	rg.w.Writefmtln("            };")
+	rg.w.Writefmtln("            return (%s)ResourceOptions.Merge(defaultOptions, options);", optionsType)
+	rg.w.Writefmtln("        }")
+
+	// Write the `Get` method for reading instances of this resource unless this is a provider resource.
+	if !rg.res.IsProvider() {
+		rg.w.Writefmtln("        /// <summary>")
+		rg.w.Writefmtln("        /// Get an existing %s resource's state with the given name, ID, and optional extra",
+			className)
+		rg.w.Writefmtln("        /// properties used to qualify the lookup.")
+		rg.w.Writefmtln("        ///")
+		rg.w.Writefmtln("        /// <param name=\"name\">The unique name of the resulting resource.</param>")
+		rg.w.Writefmtln("        /// <param name=\"id\">The unique provider ID of the resource to lookup.</param>")
+		rg.w.Writefmtln("        /// <param name=\"state\">Any extra arguments used during the lookup.</param>")
+		rg.w.Writefmtln("        /// <param name=\"options\">A bag of options that control this resource's behavior</param>")
+		rg.w.Writefmtln("        /// </summary>")
+		rg.w.Writefmtln("        public static %s Get(string name, Input<string> id, %s? state = null, %s? options = null)",
+			className, rg.res.statet.name, optionsType)
+		rg.w.Writefmtln("        {")
+		rg.w.Writefmtln("            return new %s(name, id, state, options);", className)
+		rg.w.Writefmtln("        }")
+	}
+
+	// Close the class.
+	rg.w.Writefmtln("    }")
+}
+
+func (rg *csharpResourceGenerator) generateAlias(alias tfbridge.AliasInfo) {
+	rg.w.WriteString("new Alias { ")
+	var parts []string
+	if alias.Name != nil {
+		parts = append(parts, fmt.Sprintf("Name = \"%s\"", *alias.Name))
+	}
+	if alias.Project != nil {
+		parts = append(parts, fmt.Sprintf("Project = \"%s\"", *alias.Project))
+	}
+	if alias.Type != nil {
+		parts = append(parts, fmt.Sprintf("Type = \"%s\"", *alias.Type))
+	}
+	rg.w.WriteString(strings.Join(parts, ", "))
+	rg.w.WriteString(" }")
+}
+
+func (rg *csharpResourceGenerator) generateResourceArgs() {
+	rg.generateInputType(rg.res.argst, "Pulumi.ResourceArgs", false /*nested*/, false /*plainType*/)
+}
+
+func (rg *csharpResourceGenerator) generateResourceState() {
+	if !rg.res.IsProvider() {
+		rg.generateInputType(rg.res.statet, "Pulumi.ResourceArgs", false /*nested*/, false /*plainType*/)
+	}
+}
+
+func (rg *csharpResourceGenerator) generateDatasourceFunc() {
+	// Open the partial class we'll use for datasources.
+	// TODO(pdg): this needs a better name that is guaranteed to be unique.
+	rg.w.Writefmtln("    public static partial class Invokes")
```

Comment:
do we have an official name we call "data sources" to customers?  i can't find anything in our docs....

Reply:
Can we avoid using a fixed name here? Now every namespace has `Invokes` so if I import two or more, I have to fully-qualify the names.

Reply:
What do we expect to recommend to users as the usual way to use these namespaces? e.g. will we recommend something like

```
using Ec2 = Pulumi.Aws.Ec2;
using S3 = Pulumi.Aws.S3;
```

or

```
using Pulumi.Aws.Ec2;
using Pulumi.Aws.S3;
```

?

In the former case, I think `Invokes` is fine. In the latter case, we could give these classes the same name as the containing namespace.

Reply:
In general, the latter is the default. IDEs will do those automatically. The former is a workaround when needed. We'd need a good reason to recommend aliases as the default.